src/emmc.c

```diff
@@ -0,0 +1,178 @@
+#include <linux/mmc/ioctl.h>
+#include <string.h>
+#include <sys/ioctl.h>
+#include <unistd.h>
+
+#include "emmc.h"
+#include "update_handler.h"
+
+
+static int r_emmc_read_extcsd(int fd, guint8 extcsd[512]) {
+	struct mmc_ioc_cmd cmd = {};
+
+	cmd.write_flag = 0;
+	cmd.opcode = MMC_SEND_EXT_CSD;
+	cmd.arg = 0;
+	cmd.flags = MMC_RSP_SPI_R1 | MMC_RSP_R1 | MMC_CMD_ADTC;
+	cmd.blksz = 512;
+	cmd.blocks = 1;
+	mmc_ioc_cmd_set_data(cmd, extcsd);
+
+	return ioctl(fd, MMC_IOC_CMD, &cmd);
+}
+
+gboolean r_emmc_read_bootpart(const gchar *device, gint *bootpart_active, GError **error) {
+	gint ret;
+	gboolean res = FALSE;
+	guint8 extcsd[512];
+	int fd;
+	/* count from 1 */
+	gint active_partition = -1;
+
+	fd = g_open(device, O_RDONLY);
+	if (fd == -1) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"opening eMMC device failed: %s", g_strerror(errno));
+		goto out;
+	}
+
+	ret = r_emmc_read_extcsd(fd, extcsd);
+	if (ret) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Could not read from extcsd register %d in %s\n",
+				EXT_CSD_PART_CONFIG, device);
+		goto out;
+	}
+
+	/* retrieve active partition from BOOT_PART register */
+	active_partition = (extcsd[EXT_CSD_PART_CONFIG] & 0x38) >> 3;
+
+	/* return the partitions counted from 0 */
+	*bootpart_active = active_partition - 1;
+
+	res = TRUE;
+
+out:
+	if (fd >= 0)
+		g_close(fd, NULL);
+
+	return res;
+}
+
+static gint r_emmc_write_extcsd(int fd, guint8 index, guint8 value) {
+	struct mmc_ioc_cmd cmd;
+
+	memset(&cmd, 0, sizeof(cmd));
+
+	cmd.write_flag = 1;
+	cmd.opcode = MMC_SWITCH;
+	cmd.arg = (MMC_SWITCH_MODE_WRITE_BYTE << 24) | (index << 16) |
+	          (value << 8) | EXT_CSD_CMD_SET_NORMAL;
+	cmd.flags = MMC_RSP_SPI_R1B | MMC_RSP_R1B | MMC_CMD_AC;
+
+	return ioctl(fd, MMC_IOC_CMD, &cmd);
+}
+
+gboolean r_emmc_write_bootpart(const gchar *device, gint bootpart_active, GError **error) {
+	gint ret;
+	gboolean res = FALSE;
+	int fd;
+	guint8 value = 0;
+
+	g_assert(bootpart_active == 0 || bootpart_active == 1);
+
+	fd = g_open(device, O_RDWR);
+	if (fd == -1) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"opening eMMC device failed: %s", g_strerror(errno));
+		goto out;
+	}
+
+	/* write [5:3] : BOOT_PARTITION_ENABLE of PARTITION_CONFIG */
+	if (bootpart_active == 0)
+		value = 0x08;
+	else if (bootpart_active == 1)
+		value = 0x10;
+
+	ret = r_emmc_write_extcsd(fd, EXT_CSD_PART_CONFIG, value);
+	if (ret) {
+		g_set_error(error, R_UPDATE_ERROR, R_UPDATE_ERROR_FAILED,
+				"Could not write 0x%02x to extcsd register %d in %s\n",
+				value, EXT_CSD_PART_CONFIG, device);
+		goto out;
+	}
+
+	res = TRUE;
+
+out:
+	if (fd >= 0)
+		g_close(fd, NULL);
+
+	return res;
+}
+
+static gboolean r_emmc_force_part_write(const gchar *device, gchar value, GError **error) {
+	gboolean ret = FALSE;
+	gchar *device_basename = g_path_get_basename(device);
+	gchar *sysfs_path = NULL;
+	FILE *f = NULL;
+
+	g_assert(value == '0' || value == '1');
```

Comment:
Please use `g_return_{val_,}if_fail()` for checking input args of functions. Also consider checking other inputs, such as GError, too.

Reply:
I see the point. Is this something you are trying to enforce everywhere now? There are different variants of g_assert used for argument checking in the very same file, e.g. [here](https://github.com/rauc/rauc/blob/master/src/update_handler.c#L557).

Reply:
Yes, we started with `g_assert` in some points but we would like to stick to glib convention and use g_return.. functions here. We tried to enforce this in the last patches added (while not being sure if we succeeded). The intention is to fix this with all newly touched code and later on also in existing code. Anyway, there a still a bunch of functions where we don't do any checks at all. Good patch opportunities.. ;)